hedera-node/hedera-app/src/main/java/com/hedera/node/app/ServicesMain.java

```diff
@@ -321,6 +353,118 @@ public static void main(final String... args) throws Exception {
+                        current.roundNumber(),
+                        rosterStore.get(previous.activeRosterHash()),
+                        previous.roundNumber());
+            } else {
+                // If there is no roster state content, this is a fatal error: The migration did not happen on software
+                // upgrade.
+                throw new IllegalStateException("No active rosters found in the roster state");
+            }
+        }
+
+        // Migration Software Upgrade
+        // The roster state is empty (no candidate roster and no active rosters)
+        final var activeRoster = rosterStore.getActiveRoster();
+        final var candidateRoster = rosterStore.getCandidateRoster();
+        if (activeRoster == null && candidateRoster == null) {
+            // Read the current AddressBooks from the platform state.
+            // previousRoster := translateToRoster(currentAddressBook)
+            final var addressBookStore = new ReadableStoreFactory(state).getStore(ReadableNodeStore.class);
+            final var prevousRoster = addressBookStore.snapshotOfFutureRoster();
+
+            // configAddressBook := Read the address book in config.txt
+            // currentRoster := translateToRoster(configAddressBook)
+            final var currentRoster = createRoster(diskAddressBook);
+
+            // currentRound := state round +1
+            final long currentRound = rosterStore.getRosterHistory().getFirst().roundNumber() + 1;
+            // set (previousRoster, 0) as the active roster in the roster state.
+            // set (currentRoster, currentRound) as the active roster in the roster state.
+            rosterStore.putActiveRoster(prevousRoster, 0);
+            rosterStore.putActiveRoster(currentRoster, currentRound);
+
+            // rosterHistory := [(currentRoster, currentRound), (previousRoster, 0)]
+            return new RosterHistory(currentRoster, currentRound, prevousRoster, 0);
+        }
+
+        // Subsequent Software Upgrades
+        // There is a candidate roster in the roster state
+        // No override-config.txt file is present on disk
+        if (candidateRoster != null && !overrideConfigExists) {
+            final var previousRoundPair = rosterStore.getRosterHistory().getFirst();
+
+            // currentRound := state round +1
+            final long currentRound = previousRoundPair.roundNumber() + 1;
+
+            // (previousRoster, previousRound) := read the latest (current) active roster and round from the roster
+            // state.
+            final var previousRoster = rosterStore.getActiveRoster();
+            final var previousRound = previousRoundPair.roundNumber();
+
+            // clear the candidate roster from the roster state.
+            // set (candidateRoster, currentRound) as the new active roster in the roster state.
+            rosterStore.putActiveRoster(candidateRoster, currentRound);
```

Comment:
I don't see any checks to verify that the candidate roster is ready for adoption (i.e. is fully keyed)

Reply:
I had such logic in a previous commit [here](https://github.com/hashgraph/hedera-services/pull/16082/commits/1877b6e478baaa89d690fd321c992e7f56985e10#diff-cb2dd06fe84be72eddc045d4063a3b3deb9cfbb550dc0398f470e884f275ac7bR384-R395), but then https://github.com/hashgraph/hedera-services/pull/16067 happened and I updated my PR according to it.

One question I had in that proposal PR was ([here](https://github.com/hashgraph/hedera-services/pull/16067#discussion_r1827830195)) what do we mean by "unconditionally" when we say the following:
`On software upgrade, during normal execution, if a candidate roster is present in the state, it will be unconditionally adopted as the current or latest active roster in the new roster history.`

There I mentioned if we should check if we can recreate the ledger id before adopting but checking if the roster is fully keyed is an additional condition.

I am fine with bringing checks to see if candidate roster is ready for adoption, just trying to understand when do we need them.

Reply:
@MiroslavGatsanoga   The issue this PR closes is from `TSS-Ledger-ID` epic.   `TSS-Roster` is a separate epic of work that needs to be completed before the `TSS-Ledger-ID` issue can be worked.     The issues in `TSS-Roster` that deals with startup are these: 
1. https://github.com/hashgraph/hedera-services/issues/16355
1. https://github.com/hashgraph/hedera-services/issues/16357  

`TSS-Roster` epic: https://github.com/hashgraph/hedera-services/issues/14706

Reply:
Thanks @edward-swirldslabs, I assigned myself on https://github.com/hashgraph/hedera-services/issues/16355 and will update this PR to have the implementation for it. Will mark https://github.com/hashgraph/hedera-services/issues/14751 as blocked for now, until we have the `TSS-Roster` epic done.

Reply:
@poulok Added `hasEnoughKeyMaterial` candidate roster check in https://github.com/hashgraph/hedera-services/pull/16082/commits/709ebe91e7ae73b0e9874241d255931f735bb5ca.

Reply:
The title of the PR needs to be changed.

Reply:
Removed all code that modifies state, that will be done in https://github.com/hashgraph/hedera-services/issues/16552.